src/functions.py

```diff
@@ -0,0 +1,120 @@
+"""The main script file for Pyodide."""
+
+from js import Event, document
+from pyodide.ffi import create_proxy
+from pyodide.http import pyfetch
+
+from frontend import CLEAR_BUTTON, EXECUTE_BUTTON, clear_interface, update_table
+from parser import ParentKind, TokenKind, Tree, parse, tokenize
+
+
+def flatten_response(data: dict, query: list | None = None) -> dict:
+    """Flatten a dictionary."""
+    if query is None:
+        query = []
+    flattened_result = {}
+
+    def _flatten(current: dict, name: str = "") -> dict:
+        if isinstance(current, dict):
+            for field, value in current.items():
+                _flatten(value, name + field + "_")
+        elif isinstance(current, list):
+            """old code
+            # for idx, i in enumerate(current):
+            #     _flatten(i, name + str(idx) + "_")
+            """
+        elif query and name[:-1] in query:
+            flattened_result[name[:-1]] = current  # Drops the extra _
+        else:
+            flattened_result[name[:-1]] = current
+
+    _flatten(data)
+    return flattened_result
+
+
+def extract_actor(tree: Tree) -> str | None:
+    """Extract the actor from the tree."""
+    if not tree.kind == ParentKind.FILE:
+        raise ValueError
+    stmt = tree.children[0]
+    for c in stmt.children:
+        if c.kind == ParentKind.WHERE_CLAUSE:
+            where_clause = c
+            break
+    else:
+        return None
+
+    expr = where_clause.children[1]
+
+    if expr.kind != ParentKind.EXPR_BINARY or expr.children[1].kind != TokenKind.EQUALS:
+        return None
+
+    if (
+        expr.children[0].children[0].kind == TokenKind.IDENTIFIER
+        and expr.children[0].children[0].text == "actor"
+        and expr.children[2].children[0].kind == TokenKind.STRING
+    ):
+        v = expr.children[2].children[0].text
+        return v[1:-1] if v[0] == "'" else v
+    if (
+        expr.children[2].children[0].kind == TokenKind.IDENTIFIER
+        and expr.children[2].children[0].text == "actor"
+        and expr.children[0].children[0].kind == TokenKind.STRING
+    ):
+        v = expr.children[0].children[0].text
+        return v[1:-1] if v[0] == "'" else v
+    return None
+
+
+def extract_fields(tree: Tree) -> str | None:
+    """Extract the actor from the tree."""
+    if not tree.kind == ParentKind.FILE:
+        raise ValueError
+    stmt = tree.children[0]
+    for c in stmt.children:
+        if c.kind == ParentKind.FIELD_LIST:
```

Comment:
Field list is guaranteed to be index [1], since we are assuming there's no errors in the query